test/matrix.lean

```diff
@@ -0,0 +1,143 @@
+/-
+manually ported from
+https://github.com/leanprover-community/mathlib/blob/4f4a1c875d0baa92ab5d92f3fb1bb258ad9f3e5b/test/matrix.lean
+-/
+import Mathlib.Data.Matrix.Notation
+-- import linear_algebra.Matrix.determinant
+import Mathlib.GroupTheory.Perm.Fin
+-- import Mathlib.Tactic.NormSwap
+import Qq
+
+open Qq
+
+-- todo: uncomment above imports when they are ported
+
+variables {α β : Type} [Semiring α] [Ring β]
+
+namespace Matrix
+
+open Matrix
+
+/-! Test that the dimensions are inferred correctly, even for empty matrices -/
+section dimensions
+
+-- set_option pp.universes true
+-- set_option pp.all true
+
+open Lean Meta Elab in
+def getDims (me : TermElabM Term) : TermElabM (Q(Nat) × Q(Nat)) := do
+  let e ← me
+  let elem_t ← mkFreshExprMVar (mkSort levelOne)
+  let m ← mkFreshExprMVar (mkConst ``Nat)
+  let n ← mkFreshExprMVar (mkConst ``Nat)
+  let matrix_t := mkAppN (mkConst ``Matrix [levelZero, levelZero, levelZero])
+                    #[mkApp (mkConst ``Fin) m, mkApp (mkConst ``Fin) n, elem_t]
+  let _ ← Elab.Term.elabTerm e (some matrix_t)
+  Elab.Term.synthesizeSyntheticMVarsUsingDefault
+  let m ← instantiateMVars m
+  let n ← instantiateMVars n
+  if m.hasMVar || n.hasMVar then
+    throwError "indices have metavariables"
+  return (m, n)
+
+open Lean Elab in
+def run (x : TermElabM α) : Command.CommandElabM α := Command.liftTermElabM x
+
+-- we test equality of expressions here to ensure that we have `2` and not `1.succ` in the type
+run_cmd run do let d ← getDims ``(!![]);        guard $ d == (q(0), q(0))
```

Comment:
I'm not completely sure what double backquote is, but I'm under the impression it's a version of quoting where it checks that every identifier refers to something. Single backquotes are probably more appropriate here.
```suggestion
run_cmd run do let d ← getDims `(!![]);        guard $ d == (q(0), q(0))
```
(Lean 3 was pexprs vs exprs, right?)

Reply:
Done